ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/PrimitivePoolExtensions.kt

```diff
@@ -0,0 +1,175 @@
+
+package io.kinference.ndarray.extensions
+
+import io.kinference.ndarray.arrays.*
+import io.kinference.ndarray.arrays.pointers.LongPointer
+import io.kinference.ndarray.arrays.pointers.PrimitivePointer
+import io.kinference.ndarray.extensions.utils.*
+import io.kinference.primitives.annotations.*
+import io.kinference.primitives.types.*
+
+fun PrimitiveNDArray.maxPool(
+    kernel: IntArray,
+    pads: IntArray,
+    strides: IntArray,
+    dilations: IntArray,
+    ceilMode: Int,
+    storageOrder: Int,
+    minValue: PrimitiveType = PrimitiveType.MIN_VALUE
+): List<NumberNDArrayCore> {
+    val xShapeWithPads = getShapeWithPads(shape, pads)
+    val kShapeWithDilations = IntArray(kernel.size + 2) { i -> if (i < 2) 1 else (kernel[i - 2] - 1) * dilations[i - 2] + 1 }
+
+    val resultShape = IntArray(shape.size) {
+        when (it) {
+            0, 1 -> shape[it]
+            else -> {
+                if (ceilMode == 1)
+                    ((xShapeWithPads[it] - kShapeWithDilations[it]) divCeil strides[it - 2]) + 1
+                else
+                    ((xShapeWithPads[it] - kShapeWithDilations[it]) / strides[it - 2]) + 1
+            }
+        }
+    }
+
+    if (ceilMode == 1) {
+        for (i in kernel.size until 2 * kernel.size)
+            pads[i]++
+    }
+
+    val result = MutablePrimitiveNDArray(resultShape)
+
+    val xShrankShape = shape - kShapeWithDilations + IntArray(shape.size) { 1 }
+    val xIterator = PrimitiveTensorIterator(this, xShrankShape, pads, strides)
+
+    val w = PrimitiveNDArray(IntArray(kernel.size + 2) { i -> if (i < 2) 1 else kernel[i - 2] })
+    val wIterator = PrimitiveTensorIterator(w, w.shape)
+    val wSize = calculateInnerShapeSize(w.shape)
+    val rawShifts = IntArray(wSize)
+    val indexShifts = Array(wSize) {
+        val cur = wIterator.next() * dilations
+        rawShifts[it] = calculateInnerShift(shape, cur)
+        cur
+    }
+
+    val indices = if (storageOrder == -1) MutableLongNDArray.scalar(0) else MutableLongNDArray(resultShape)
+
+    val resultPointer = PrimitivePointer(result.array)
+    val indicesPointer = LongPointer(indices.array)
+    while (xIterator.hasNext()) {
+        var res: PrimitiveType = minValue
+        xIterator.next()
+        indexShifts.forEachIndexed { i: Int, index: IntArray ->
+            val cur = xIterator.getShifted(index, rawShifts[i], minValue)
+            if (res < cur) {
+                res = cur
+                when (storageOrder) {
+                    0 -> indicesPointer.set((xIterator.rawCurrent + rawShifts[i]).toLong())
+                    1 -> indicesPointer.set(calcRawIndexWithColMajor(xIterator.current, shape) + calcRawIndexWithColMajor(index, shape))
+                }
+            }
+        }
+        indicesPointer.increment()
+        resultPointer.setAndIncrement(res)
+    }
+
+    return listOf(result, indices)
+}
+
+@SpecifyPrimitives(include = [])
+fun calcRawIndexWithColMajor(index: IntArray, shape: IntArray): Long {
```

Comment:
Move this and other functions that don't require any actual codegen to another place